lib/addons/library.kodi.audioengine/libKODI_audioengine.cpp

```diff
@@ -0,0 +1,236 @@
+ *      http://kodi.tv
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with KODI; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+#include <string>
+#include "../../../addons/library.kodi.audioengine/libKODI_audioengine.h"
+#include "addons/AddonCallbacks.h"
+
+#ifdef _WIN32
+#include <windows.h>
+#define DLLEXPORT __declspec(dllexport)
+#else
+#define DLLEXPORT
+#endif
+
+using namespace std;
+
+#define LIBRARY_NAME "libKODI_audioengine"
+
+extern "C"
+{
+
+DLLEXPORT void* AudioEngine_register_me(void *hdl)
+{
+  CB_AudioEngineLib *cb = NULL;
+  if (!hdl)
+    fprintf(stderr, "%s-ERROR: AudioEngine_register_me is called with NULL handle !!!\n", LIBRARY_NAME);
+  else
+  {
+    cb = ((AddonCB*)hdl)->AudioEngineLib_RegisterMe(((AddonCB*)hdl)->addonData);
+    if (!cb)
+      fprintf(stderr, "%s-ERROR: AudioEngine_register_me can't get callback table from KODI !!!\n", LIBRARY_NAME);
+  }
+  return cb;
+}
+
+DLLEXPORT void AudioEngine_unregister_me(void *hdl, void* cb)
+{
+  if (hdl && cb)
+    ((AddonCB*)hdl)->AudioEngineLib_UnRegisterMe(((AddonCB*)hdl)->addonData, (CB_AudioEngineLib*)cb);
+}
+
+// ---------------------------------------------
+// CAddonAEStream implementations
+// ---------------------------------------------
+DLLEXPORT CAddonAEStream* AudioEngine_make_stream(void *hdl, void *cb, AEDataFormat DataFormat, unsigned int SampleRate, unsigned int EncodedSampleRate, enum AEChannel *Channels, unsigned int Options)
+{
+  AEStreamHandle *streamHandle = ((CB_AudioEngineLib*)cb)->MakeStream(DataFormat, SampleRate, EncodedSampleRate, Channels, Options);
+  if(!streamHandle)
+  {
+    fprintf(stderr, "%s-ERROR: AudioEngine_make_stream can't MakeStrean failed!\n", LIBRARY_NAME);
+    return NULL;
+  }
+
+  return new CAddonAEStream(hdl, cb, streamHandle);
+}
+
+DLLEXPORT void AudioEngine_free_stream(CAddonAEStream *p)
+{
+  if(p)
+  {
+    delete p;
+  }
+}
+
+DLLEXPORT bool AudioEngine_get_current_sink_Format(void *hdl, void *cb, AudioEngineFormat *SinkFormat)
+{
+  if(!cb)
+  {
+    return false;
+  }
+    
+  return ((CB_AudioEngineLib*)cb)->GetCurrentSinkFormat(((AddonCB*)hdl)->addonData, SinkFormat);
+}
+
+CAddonAEStream::CAddonAEStream(void *Addon, void *Callbacks, AEStreamHandle *StreamHandle)
+{
+  if(!Addon || !Callbacks || !StreamHandle)
+  {
+    // ToDo: throw error!
```

Comment:
This is still a to do, but at the moment I don't know if it is necessary. If not I can remove it.

I wanted to throw an exception at this line.